aip/general/0236.md

```diff
@@ -0,0 +1,315 @@
+consequences which severely impact a customer’s overall infrastructure setup.
+
+Previews are a rollout safety mechanism for Policy resources which gives the
+customer the ability to validate the effect of their proposed changes against
+live production traffic prior to the changes going live.
+
+## Guidance
+
+### System changes
+
+The API logic of this mechanism pushes an experimental version of a proposed
+change to the upstream system (data plane) while the existing live version
+remains active. Both versions of the configuration are subsequently evaluated
+for incoming traffic, however only the effective version’s configurations
+actually have the ability to alter traffic. Meanwhile, the output of the
+experimental version is logged so a customer can validate that the proposed
+changes are as intended. This gives the customer the ability to fully understand
+the effects of the new configuration against live traffic before replacing
+the effective configuration with it.
+
+### Non-goals
+
+#### Non-policy resources
+
+This rollout safety mechanism is for policies only.  
+
+A Policy is a customer-owned resource with a set of rules that govern the
+behavior of how other resources admit or deny a target entity  Generally, the
+outcome of a Policy can be evaluated to a specific set of outcomes. For example,
+a FirewallPolicy evaluates to either ALLOW or DENY for incoming traffic into
+a Virtual Network. This lends well to preview mode as a rollout safety mechanism
+as the proposed changes to a Policy can simply be logged instead of enforced.
+
+### Nested collection called experiments
+
+A new configuration of a Policy which is to be validated using preview mode is
+stored as a nested collection under that Policy resource.
+
+In this AIP, we will use the example of a hypothetical Policy resource called,
+`Policy`. The following is the pattern for the resource name:
+
+`projects/{project}/locations/{location}/policies/{policy}`
+
+The experimental versions of the resource used for preview mode are stored as a
+nested collection under `Policy`. The resource type will follow the convention
+"<resource_name> + Experiment”. The following is the pattern for the experiment:
+
+`/projects/{project}/locations/{location}/policies/{policy}/experiments/{experiment}`
+
+A proto used to represent an experiment **must** contain the following fields:
+
+  1. The required top-level fields for a resource, like `name` and `etag`.
+  2. The Policy that is being tested itself.
+  3. The `preview_metadata` field for preview.
+
+
+```proto
+message PolicyExperiment {
+
+  // The name of the PolicyExperiment.
+  string name = 1;
+ 
+  // The fields used for log evaluation and eventual promotion.
+  Policy policy = 2;
+
+  // Metadata for preview mode only.
+  PreviewMetadata preview_metadata = 3 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+- A new proto **must** be used for experiments with the Policy proto inlined
```

Comment:
- A new proto **must** be used for each experiment. There **must not** be a common `<ResourceName>Preview` proto for all experiments.
    - The experiment proto **must** contain a field of the type that is the same as the policy resource to which the  experiment applies.
    - To introduce experimental behavior, change the values of this proto.